Graph split event tracker

haowu14 opened the pull request:
Summary:
A tool to track events in graph split, specifically on how nodes being end up in acc or cpu subgraphs.

Usage: use env var to specify a mode and necessary arguments.

FX_NET_ACC_SPLITTER_TRACKER_MODE: Tracker mode. 
```
Different modes of the event tracker:
"0": Tracker not enabled (by default)
"1": Tracker enabled but no dumps. Information available by setting breakpoints and visually inspect in pdb.
"2": Tracker enabled and dumps all events to DUMP_PREFIX_all.txt
"3": In addition to events dump, track nodes specified by ENV_FX_NET_ACC_SPLITTER_TRACKER_TRACKED_NODES recusrively and dump to DUMP_PREFIX_nodex.txt
"4:: In addition to events dump, track all nodes with more than 1 event recusrively and dump to DUMP_PREFIX_nodex.txt
```
FX_NET_ACC_SPLITTER_TRACKER_DUMP_PATH: overriding dump path. Leave empty for `~`.
FX_NET_ACC_SPLITTER_TRACKER_TRACKED_NODES: Nodes to track for mode "3".

Test Plan: New unit test

Reviewed By: georgiaphillips

Differential Revision: D79203595




cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv

ezyang commented on torch/fx/passes/splitter_base.py:
```diff
@@ -39,6 +40,35 @@
+"""
+TRACKER_MODE: Literal["0", "1", "2", "3", "4"] = os.environ.get(
+    ENV_FX_NET_ACC_SPLITTER_TRACKER_MODE, "0"
+)  # type: ignore[assignment]
```
Instead of doing it by an envvar, can we just dump it by default to tlparse using trace_structured?

haowu14 replied:
I can make the full flat event dump by default and always dump to the path specified by TORCH_TRACE. But I want a way for the user to specify the nodes they are interested in and dump those nodes recursively. So seems we can't get rid of all env vars anyway?

haowu14 replied:
I think I can make a default dump of all events flat via trace_structured, while preserving the local dump for a quick debug process. 
Later we can followup to visualize in fancier ways via tlparse.

ezyang commented on test/fx/test_fx_split_node_finder.py:
```diff
@@ -0,0 +1,79 @@
+
+        # override tracker without having to run with env var
+        tracker = NodeEventTracker(
+            1,  # mode: just enable the tracker without dumping
```
You should test the 2 and 3 modes too?

ezyang commented on torch/fx/passes/splitter_base.py:
```diff
@@ -99,6 +133,141 @@ def __init__(
+                "name": "fx_net_acc_splitter_finder_events",
+                "encoding": "json",
+            },
+            payload_fn=lambda: json.dumps(self.to_dict()),
```
Help me understand how big this is, typically?

haowu14 replied:
In a graph with 12k nodes, this is around 4MB.

ezyang commented on torch/fx/passes/splitter_base.py:
```diff
@@ -186,17 +363,22 @@ def reduce_acc_nodes_non_tensor_output(self):
             self.reduce_acc_nodes_non_tensor_input()
             self.reduce_acc_nodes_non_tensor_output()
-
+        self.tracker.close()
```
is it ok not to close if an exception is raised here?

ezyang reviewed (approved):
I'm going to accept this in the spirit of not blocking things, but in general it is good to figure out how to not add lots off one off env vars and go through the preexisting mechanisms for improved discoverability. For example, consider if the log info here can simply be exposed with TORCH_LOGS in tandem with trace_structured, and avoid writing out files entirely. I'm not blocking it because we have several flags that write out files but in general no one except you is going to know that you can do this.  Something that pops up in tlparse has much more chance of being found by others.

haowu14 committed:
Graph split event tracker (#159795)

Summary:

A tool to track events in graph split, specifically on how nodes being end up in acc or cpu subgraphs.

Usage: use env var to specify a mode and necessary arguments.

FX_NET_ACC_SPLITTER_TRACKER_MODE: Tracker mode. 
```
Different modes of the event tracker:
"0": Tracker enabled but no local dumps. Information available by setting breakpoints and visually inspect in pdb.
"1": Tracker enabled and dumps all events to DUMP_PREFIX_all.txt
"2": In addition to events dump, track nodes specified by ENV_FX_NET_ACC_SPLITTER_TRACKER_TRACKED_NODES
     recursively and dump to DUMP_PREFIX_nodex.txt
"3": In addition to events dump, track all nodes with more than 1 event recursively and dump to DUMP_PREFIX_nodex.txt
Regardless of the modes, tracker is always enabled and dumps by trace_structured.
```
FX_NET_ACC_SPLITTER_TRACKER_DUMP_PATH: overriding dump path. Leave empty for `~`.
FX_NET_ACC_SPLITTER_TRACKER_TRACKED_NODES: Nodes to track for mode "2".

Test Plan:
New unit test

```
buck test caffe2/test:fx -- test_fx_split_node_finder
```


---- 


```
TORCH_TRACE=~/my_trace_log_dir FX_NET_ACC_SPLITTER_TRACKER_MODE=3 ../buck-out/v2/gen/fbcode/6f6fe98d41631b2e/inference_enablement/model_processing/infra/components/lowering/re/__re_cinder__/re_cinder.par -r '{"aot_inductor":{"serialized_inference_model_input_path":"ads_storage_fblearner/tree/user/facebook/fblearner/predictor/895540436/4/gpu_lowering/input.merge.61759375","serialized_inference_model_output_path":"ads_storage_fblearner/tree/user/facebook/fblearner/predictor/895540436/4/gpu_lowering/inductor_output.merge.61759375","submodule_names_to_lower":["merge"],"inductor_lowering_context":{"aot_inductor_lowering_settings":{"max_batch_size":2048,"min_acc_module_size":10,"workdir":"/tmp/local","name":"merge","dll_name":"inductor_engine.so","use_scripting":true,"preset_lowerer":"gr;disable_new_lowering_weights;disable_dper_passes:passes=fuse_parallel_linear_no_weight_change|fuse_parallel_linear","precision":4,"output_precision":4,"remote_cache_file_path_folder":"ads_storage_fblearner/tree/user/facebook/fblearner/predictor/895540436/","save_remote_cache":true,"aot_inductor_config":"{\"max_autotune\":True,\"comprehensive_padding\":False}","disable_dynamic_shapes":false,"remove_unexpected_type_cast":false,"disable_constraint_solver":false,"sample_input_tile_factor":32,"disable_acc_tracer":true,"generate_sample_inputs":true,"tile_sample_input_by_dynamic_shape":false,"node_replacement_dict":"","dynamic_shapes_strategy":73728,"auto_dynamic_shapes":false,"auto_dynamic_shapes_min_size":1,"auto_dynamic_shapes_max_size":1048576,"max_acc_splits":-1,"dynamic_size":-1,"pre_dispatch_export":true,"merge_split_optimization":false,"use_dynamic_dim_hints":false,"allow_refine_dynamic_shapes_on_constants":false,"use_sigmoid":false}},"model_entity_id":895540436,"model_snapshot_id":4,"add_sample_inputs":false,"platform_arch":0,"lowering_lib_pkg":"ien.lower:prod","dense_in_place_format":2}}'
```

Events dump: P1896093119
Nodes track dump:  P1896110514
The above files are generated locally
```
? _fx_net_tracker_all.txt
? _fx_net_tracker_nodes.txt
```

Also in torch_trace you can find all events: https://www.internalfb.com/intern/paste/P1897874179/

Rollback Plan:

Reviewed By: georgiaphillips, ezyang

Differential Revision: D79203595
